File: src/main/java/pl/training/workshop/m3/s02_similarity/step1/OnlineCheckout.java
package pl.training.workshop.m3.s02_similarity.step1;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

/**
 * Krok 1: Inline Method - wspólna metoda wróciła do obu wywołujących.
 * Tymczasowe powtórzenie kodu to bezpieczny etap rozdzielania reguł.
 */
public final class OnlineCheckout {
    public BigDecimal total(List<BigDecimal> ticketPrices) {
        BigDecimal tickets = ticketPrices.stream().reduce(BigDecimal.ZERO, BigDecimal::add);
        BigDecimal perUnit = ServiceFee.Kind.ONLINE_BOOKING == ServiceFee.Kind.ONLINE_BOOKING
                ? new BigDecimal("2.00") : new BigDecimal("3.00");
        return tickets.add(perUnit.multiply(BigDecimal.valueOf(ticketPrices.size()))
                .setScale(2, RoundingMode.HALF_UP));
    }
}

File: csharp/src/Training.Workshop/M3/S02Similarity/Step1/OnlineCheckout.cs
namespace Training.Workshop.M3.S02Similarity.Step1;

/// <summary>
/// Krok 1: Inline Method - wspólna metoda wróciła do obu wywołujących.
/// Tymczasowe powtórzenie kodu to bezpieczny etap rozdzielania reguł.
/// </summary>
public sealed class OnlineCheckout
{
    public decimal Total(IReadOnlyList<decimal> ticketPrices)
    {
        var tickets = ticketPrices.Sum();
#pragma warning disable CS1718 // porównanie stałej ze sobą - ślad po Inline Method, usuwany w kroku 2
        var perUnit = ServiceFee.Kind.OnlineBooking == ServiceFee.Kind.OnlineBooking
            ? 2.00m : 3.00m;
#pragma warning restore CS1718
        return tickets + Math.Round(perUnit * ticketPrices.Count, 2, MidpointRounding.AwayFromZero);
    }
}
